src/trace/persistent/cursor.rs

```diff
@@ -0,0 +1,284 @@
+        // A value is valid if `cur_vals` is set and we have not iterated past
+        // the length of the current values.
+        self.cur_vals.is_some() && self.val_idx < self.cur_vals.as_ref().unwrap().len()
+    }
+
+    #[inline]
+    fn key(&self) -> &B::Key {
+        self.cur_key.as_ref().unwrap()
+    }
+
+    #[inline]
+    fn val(&self) -> &B::Val {
+        &self.cur_vals.as_ref().unwrap()[self.val_idx].0
+    }
+
+    #[inline]
+    fn map_times<L: FnMut(&B::Time, &B::R)>(&mut self, mut logic: L) {
+        if self.val_valid() {
+            for (time, val) in self.cur_vals.as_ref().unwrap()[self.val_idx].1.iter() {
+                logic(time, val);
+            }
+        }
+    }
+
+    fn map_times_through<L: FnMut(&B::Time, &B::R)>(&mut self, mut logic: L, upper: &B::Time) {
+        if self.key_valid() && self.val_valid() {
+            // Note that `map_times_through` uses `less_equal` to determine if
+            // `logic` should be called on a given `(time, diff)`. However,
+            // `cur_vals` is sorted by `Ord` hence we have to go through all the
+            // values to determine how many times `logic` should be called.
+            for (time, val) in self.cur_vals.as_ref().unwrap()[self.val_idx].1.iter() {
+                if time.less_equal(upper) {
+                    logic(time, val);
+                }
+            }
+        }
+    }
+
+    #[inline]
+    fn weight(&mut self) -> B::R
+    where
+        B::Time: PartialEq<()>,
+    {
+        // This is super ugly because the RocksDB storage format is generic and
+        // needs to support all the different data-structures, but if we can
+        // call weight we can basically just access the first elements in the
+        // list of lists of lists of lists...
+        self.cur_vals.as_ref().unwrap()[0].1[0].1.clone()
+    }
+
+    #[inline]
+    fn step_key(&mut self) {
+        if self.db_iter.valid() {
+            // Note: RocksDB only allows to call `next` on a `valid` cursor
+            self.db_iter.next();
+
+            if self.db_iter.valid() {
+                self.update_current_key_weight();
+            } else {
+                // Reached the end of the iterator.
+                //
+                // We need to "record" that we reached the end by storing the
+                // last key in the `cursor_upper_bound`.
+                //
+                // Otherwise something like  keys = [1], operations = [StepKey,
+                // SeekKey(0)] would result in key_valid() returning true (see
+                // `seek_key`)
+                self.cursor_upper_bound = self.cur_key.take();
+                self.cur_vals = None;
+            }
+        }
+    }
+
+    /// Returns the last key in the cursor or `None` if the cursor is empty.
+    ///
+    /// # TODO
+    /// This is obviously wrong but doesn't fail any tests...
+    fn last_key(&mut self) -> Option<&B::Key> {
+        None
+    }
+
+    #[inline]
```

Comment:
I don't think we want to inline such large chunks of code.

Reply:
Fixed, removed